nucliadb_vectors2/src/vectors/data_point/ops_hnsw.rs

```diff
@@ -232,32 +233,78 @@ impl<'a, DR: DataRetriever> HnswOps<'a, DR> {
     }
     pub fn search<H: Hnsw>(
         &self,
-        x: Address,
+        query: Address,
         hnsw: H,
         k_neighbours: usize,
         with_filter: &[&[u8]],
+        with_duplicates: bool,
     ) -> Neighbours {
         if let Some(entry_point) = hnsw.get_entry_point() {
             let mut crnt_layer = entry_point.layer;
             let mut neighbours = vec![(entry_point.node, 0.)];
             while crnt_layer != 0 {
+                let layer = hnsw.get_layer(crnt_layer);
                 let entry_points: Vec<_> = neighbours.into_iter().map(|(node, _)| node).collect();
-                let layer_res = self.layer_search(x, hnsw.get_layer(crnt_layer), 1, &entry_points);
+                let layer_res = self.layer_search(query, layer, 1, &entry_points);
                 neighbours = layer_res;
                 crnt_layer -= 1;
             }
             let entry_points: Vec<_> = neighbours.into_iter().map(|(node, _)| node).collect();
             let layer = hnsw.get_layer(crnt_layer);
-            let result = self.layer_search(x, layer, k_neighbours, &entry_points);
-            let mut solution = result.iter().copied().map(|v| v.0).collect();
-            result
-                .into_iter()
-                .flat_map(|(n, _)| {
-                    self.closest_up_node(&mut solution, with_filter, hnsw.get_layer(0), n, x)
-                })
-                .collect()
+            let result = self.layer_search(query, layer, k_neighbours, &entry_points);
+            let mut sol_addresses = HashSet::new();
+            let mut vec_counter = RepCounter::new(!with_duplicates);
+            let mut filtered_result = Vec::new();
+            result.iter().copied().for_each(|(addr, score)| {
+                sol_addresses.insert(addr);
+                vec_counter.add(score);
+            });
+            result.into_iter().for_each(|(addr, score)| {
+                sol_addresses.remove(&addr);
+                vec_counter.sub(score);
+                if let Some((adr, score)) = self.closest_up_node(
```

Comment:
I think you can reuse `addr` variable name here (and shadow the previous one) instead of having a variable names way too similar

Reply:
Its a typo, will do

Reply:
fixed by [363880a](https://github.com/nuclia/nucliadb/pull/353/commits/363880a41a5053d81c5f3b42f5abbc68d8f4bd92)